tests/mlio-test/test_text_line_reader.cxx

```diff
@@ -0,0 +1,113 @@
+#include <gtest/gtest.h>
+#include <mlio.h>
+#include <unistd.h>
+
+#include "../../src/mlio/detail/error.h"
+
+namespace mlio {
+
+class test_text_line_reader : public ::testing::Test {
+protected:
+    test_text_line_reader()
+    {
+        pathname_ = "/tmp/mlio-test-text-line-reader-XXXXXX.txt";
+        fd_ = ::mkstemps(&pathname_.front(), 4);
+        if (fd_ == -1) {
+            throw std::runtime_error(
+                "Could not open file for generating input");
+        }
+        const char *mesg = "this is line 1\nthis is line 2\nthis is line 3\n";
+        write(fd_, mesg, strlen(mesg));
+    }
+
+    ~test_text_line_reader() override;
+
+protected:
+    int fd_;
+    std::string const expected_line_1_ = "this is line 1";
+    std::string const expected_line_2_ = "this is line 2";
+    std::string const expected_line_3_ = "this is line 3";
+    std::string pathname_;
+};
+
+test_text_line_reader::~test_text_line_reader()
+{
+    ::unlink(&pathname_.front());
+}
+
+TEST_F(test_text_line_reader, test_text_line_reader_happy_path)
+{
+    auto dataset = mlio::list_files(pathname_, "*.txt");
+    mlio::data_reader_params prm{dataset, 3};
+
+    auto reader = mlio::make_intrusive<mlio::text_line_reader>(prm);
+    for (auto i = 0; i < 2; i++) {
+        mlio::intrusive_ptr<mlio::example> exm;
+        while ((exm = reader->read_example()) != nullptr) {
+            auto lbl =
+                static_cast<dense_tensor *>(exm->find_feature("value").get());
+            auto str_ptr = static_cast<std::string *>(lbl->data().data());
```

Comment:
`lbl->data()` returns a value of type `device_array_span` which has a convenience member function called `as<typename>()`. You should use that instead of casting types yourself.

```
auto strings = lbl->data.as<std::string>();

EXPECT_EQ(strings[0], expected_line_1_)
EXPECT_EQ(strings[1], expected_line_2_);
```

This of course applies to other unit tests as well.

Reply:
got it, thanks! very useful.